contracts/TwabRewards.sol

```diff
@@ -0,0 +1,451 @@
+        require(
+            !_isClaimedEpoch(_user, _promotionId, _epochId),
+            "TwabRewards/rewards-already-claimed"
+        );
+
+        uint256 _rewardAmount = _calculateRewardAmount(_user, _promotionId, _epochId);
+
+        IERC20 _token = IERC20(_getPromotion(_promotionId).token);
+        _token.safeTransferFrom(address(this), _user, _rewardAmount);
+
+        _setClaimedEpoch(_claimedEpochs[_promotionId][_user], _epochId, true);
+
+        emit RewardsClaimed(_token, _rewardAmount);
+
+        return _rewardAmount;
+    }
+
+    /* ============ Internal Functions ============ */
+
+    /**
+    @notice Determine if address passed is actually a ticket.
+    @param _ticket Address to check
+   */
+    function _requireTicket(address _ticket) internal view {
+        require(address(_ticket) != address(0), "TwabRewards/ticket-not-zero-address");
+
+        (bool succeeded, bytes memory data) = address(_ticket).staticcall(
+            abi.encodePacked(ITicket(_ticket).controller.selector)
+        );
+
+        address controllerAddress;
+
+        if (data.length > 0) {
+            controllerAddress = abi.decode(data, (address));
+        }
+
+        require(succeeded && controllerAddress != address(0), "TwabRewards/invalid-ticket");
+    }
+
+    /**
+        @notice Determine if the number of epochs passed exceeds the maximum number of epochs.
+        @param _numberOfEpochs Number of epochs to check
+    */
+    function _requireEpochLimit(uint256 _numberOfEpochs) internal pure {
+        require(_numberOfEpochs < 256, "TwabRewards/exceeds-256-epochs-limit");
+    }
+
+    /**
+        @notice Get settings for a specific promotion.
+        @dev Will revert if the promotion does not exist.
+        @param _promotionId Promotion id to get settings for
+        @return Promotion settings
+     */
+    function _getPromotion(uint256 _promotionId) internal view returns (Promotion memory) {
+        return _promotions[_promotionId];
+    }
+
+    /**
+        @notice Get settings for a specific epoch.
+        @param _epochId Epoch id to get settings for
+        @param _promotion Promotion settings
+        @return Epoch settings
+     */
+    function _getEpoch(uint256 _epochId, Promotion memory _promotion)
+        internal
+        pure
+        returns (Epoch memory)
+    {
+        uint256 _epochDuration = _promotion.epochDuration;
+
+        return
+            Epoch({
+                id: uint32(_epochId),
+                startTimestamp: uint32(_promotion.startTimestamp + (_epochDuration * _epochId)),
+                duration: uint32(_epochDuration)
+            });
+    }
+
+    /**
+        @notice Get current epoch settings.
+        @param _promotionId Id of the promotion to get current epoch for
+        @return Epoch settings
+     */
+    function _getCurrentEpoch(uint256 _promotionId) internal view returns (Epoch memory) {
```

Comment:
Strip it down, no struct: `_getCurrentEpochId`

Reply:
Done: https://github.com/pooltogether/v4-periphery/pull/8/commits/423f188e97b1edd4ff89b6b5621bde3f656b0fbc